feat(monitor): init

noaccOS opened the pull request:
initialize process monitoring, by calling ~`Monitor.monitor`~ `TerminationCallbacks.add` with a pid and a callback

depends on #17 and includes its commits

noaccOS committed:
feat(monitor): init

initialize process monitoring, by calling `TerminationCallbacks.add`
with a pid and a callback

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

rbino commented on test/edgehog_device_forwarder/monitor_test.exs:
```diff
@@ -0,0 +1,29 @@
+  def until_exit_process do
+    receive do
+      :exit -> :ok
+      _ -> until_exit_process()
```
Receive blocks if it doesn't find any messages matching the pattern, so you can remove this and avoid the process spinning endlessly.
Also, I'd call this `wait_for_exit`

noaccOS replied:
But I want it to block, as I need to control _when_ the process has to finish

Edit: I had misunderstood the meaning of your comment, fixed it in b44d6090a4b03562c2c6d5dc3a94002a75d1d1e1

rbino commented on lib/edgehog_device_forwarder/monitor.ex:
```diff
@@ -0,0 +1,35 @@
+  @spec monitor(pid, (-> any)) :: :ok
+  def monitor(pid, on_close) when is_function(on_close, 0) do
+    GenServer.cast(__MODULE__, {:monitor, pid, on_close})
+  end
```
`Monitor` (both ad module name and function name) is a little generic and doesn't convey what this actually does. I'd call this something like `TerminationHandler` and `clean_up_with` (or propose some other names)

noaccOS replied:
I propose `TerminationCallbacks` and `add`

rbino commented on test/support/cache_case.ex:
```diff
@@ -0,0 +1,19 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.CacheCase do
```
This is unrelevant if this is the first PR that has to be merged, move this to the Cache PR

Annopaolo commented on test/edgehog_device_forwarder/termination_callbacks_test.exs:
```diff
@@ -0,0 +1,33 @@
+      :exit -> :ok
+      _ -> wait_for_exit()
+    end
+  end
```
Up what @rbino said, you don't need to recurse here. If you only put the `:exit -> :ok` branch in `receive`, `process` will block until `:exit` is received. Then, it will return `:ok` and exit. It seems that this is what you need

Annopaolo commented on lib/edgehog_device_forwarder/termination_callbacks.ex:
```diff
@@ -0,0 +1,36 @@
+  def handle_cast({:add, pid, on_close}, processes)
+      when is_function(on_close, 0) do
+    ref = Process.monitor(pid)
+    {:noreply, Map.put(processes, ref, on_close)}
```
This GenServer sttae is not persisted, meaning that the ref <-> callback map will be lost if the `TerminationCallbacks` server goes down. Is this the intention, i.e. callback might not be executed for some processes?

noaccOS commented:
I have moved the con_cache dependency here from #6 

> @Annopaolo This GenServer sttae is not persisted, meaning that the ref <-> callback map will be lost if the `TerminationCallbacks` server goes down

Now `ConCache` in used instead of a simple map to keep track of the monitored processes, allowing for the state to be restore after a crash.

This of course implies that the problem has been moved on to the `ConCache` and `:ets` process, but I think this is a reasonable compromise as other parts of the application (such as the phoenix endpoint) rely on `:ets`

Annopaolo commented on lib/edgehog_device_forwarder/termination_callbacks.ex:
```diff
@@ -0,0 +1,62 @@
+    ConCache.ets(@table)
+    |> :ets.tab2list()
+    |> Enum.map(fn {pid, _} -> pid end)
+    |> Enum.each(&Process.monitor/1)
```
This seems a good fit for a for comprehension:
```suggestion
    table = ConCache.ets(@table)
    
    for {pid, _ } <- :ets.tab2list(table) do
      Process.monitor(pid)
    end
```

Annopaolo commented on lib/edgehog_device_forwarder/termination_callbacks.ex:
```diff
@@ -0,0 +1,62 @@
+    callback =
+      ConCache.isolated(@table, pid, fn ->
+        callback = ConCache.get(@table, pid)
+        ConCache.dirty_delete(@table, pid)
```
This does not guarantee that the callback is executed before the entry is deleted from the table. 
On the other hand, executing the callback in the `ConCache.isolated` context is a no-go.


A possibility would be to split cache read from entry deletion:
1. You read the entry from cache
2. You run the callback
3. You delete the cached entry

However, this too  would not give a transactional guarantee. But, this would allow to add another process that checks if a callback for a dead process has been executed and possibly repeat it, then garbage collect dead processes removing them from the cache. I'm not sure that spinning up all of this is what we want, tough.

Another possibility is spawning a task that executes the callback in the `ConCache.isolated` context. This way you'd have the soft guarantee that the callback has at least been scheduled for execution before the related entry has been deleted:
```elixir
      ConCache.isolated(@table, pid, fn ->
        callback = ConCache.get(@table, pid)
        unless callback == nil, do: Task.start(callback/0)
        ConCache.dirty_delete(@table, pid)
      end)
```

noaccOS replied:
For our purpose, I think it's safe to use the first approach you mentioned in its simple form (ie, running the callback before deleting it)

This would guarantee that the callback is run at least one time before it gets deleted from the cache.
Right now, this works fine with the callbacks we use.

We can always complicate the structure if future problems require so.

noaccOS commented:
EDIT: looks like it works even without :transient, and using :transient didn't restart TerminationCallbacks if a callback made it crash. Whoopsie.

~TerminationCallbacks and its cache now have the `restart: transient` option~

~This is useful because it allows us to safely run the `Supervisor.terminate_child` and `Supervisor.restart_child` in our tests.~
This approach resolved an issue there was calling the `delete` on all the items where sometimes the delete was executed after the new value was inserted, resulting in some tests randomly failing

~Should not affect runtime as we never explicitly terminate these processes gracefully~

noaccOS committed:
chore: add con_cache dependency

ConCache seem to be the most adequate solution to the problem,
as it is mostly an ETS wrapper which allows for
atomic row-based operations and is designed for parallelism.

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS commented:
style commits moved to #17

rbino commented on lib/edgehog_device_forwarder/termination_callbacks.ex:
```diff
@@ -0,0 +1,59 @@
+
+  @spec add(pid, (-> any)) :: :ok
+  def add(pid, on_close) when is_pid(pid) and is_function(on_close, 0) do
+    GenServer.cast(__MODULE__, {:add, pid, on_close})
```
This should be a `call`, otherwise you don't have any guarantee about the fact the message actually went through (`cast` will happily return `:ok` if the process you're sending to is dead)

rbino commented:
Please rebase on `main`, I still see changes in `mix.exs`

noaccOS commented:
> Please rebase on `main`, I still see changes in `mix.exs`

Rebased.
I've also added a supervisor for the cache callbacks and its cache, so that if the cache dies then `TerminationCallbacks` dies as well (most useful in tests but it represents how it should normally work too)
